src/tactic/equiv_rw.lean

```diff
@@ -0,0 +1,260 @@
+-- (example goal: we could rewrite along an isomorphism of rings (either as `R ≅ S` or `R ≃+* S`)
+-- and turn an `x : mv_polynomial σ R` into an `x : mv_polynomial σ S`.).
+
+meta def equiv_congr_lemmas : tactic (list expr) :=
+do exprs ←
+  [
+  `equiv.of_iff,
+  -- These are functorial w.r.t equiv,
+  -- and so will be subsumed by `equiv_functor.map_equiv` in a subsequent PR.
+  -- Many more could be added, but will wait for that framework.
+  `equiv.perm_congr, `equiv.equiv_congr, `equiv.unique_congr,
+  -- The function arrow is technically a bifunctor `Typeᵒᵖ → Type → Type`,
+  -- but the pattern matcher will never see this.
+  `equiv.arrow_congr',
+  -- Allow rewriting in subtypes:
+  `equiv.subtype_equiv_of_subtype',
+  -- Allow rewriting in the first component of a sigma-type:
+  `equiv.sigma_congr_left',
+  -- Allow rewriting ∀s:
+  -- (You might think that repeated application of `equiv.forall_congr'
+  -- would handle the higher arity cases, but unfortunately unification is not clever enough.)
+  `equiv.forall₃_congr',
+  `equiv.forall₂_congr',
+  `equiv.forall_congr',
+  -- Allow rewriting in argument of Pi types:
+   `equiv.Pi_congr_left',
+  -- Handles `sum` and `prod`, and many others:
+  `bifunctor.map_equiv,
+  -- Handles `list`, `option`, and many others:
+  `functor.map_equiv,
+  -- We have to filter results to ensure we don't cheat and use exclusively `equiv.refl` and `iff.refl`!
+  `equiv.refl,
+  `iff.refl
+  ].mmap (λ n, try_core (mk_const n)),
+  return (exprs.map option.to_list).join -- TODO: implement `.mfilter_map mk_const`?
+
+declare_trace adapt_equiv
+
+/-- Implementation of `adapt_equiv`, using `solve_by_elim`. -/
+meta def adapt_equiv_core (eq ty : expr) : tactic expr :=
+do
+  -- We prepare a synthetic goal of type `(%%ty ≃ _)`, for some placeholder right hand side.
+  initial_goals ← get_goals,
+  g ← to_expr ``(%%ty ≃ _) >>= mk_meta_var,
+  set_goals [g],
+  -- Assemble the relevant lemmas.
+  equiv_congr_lemmas ← equiv_congr_lemmas,
+  /-
+    We now call `solve_by_elim` to try to generate the requested equivalence.
+    There are a few subtleties!
+    * We make sure that `eq` is the first lemma, so it is applied whenever possible.
+    * In `equiv_congr_lemmas`, we put `equiv.refl` last so it is only used when it is not possible
+      to descend further.
+    * To avoid the possibility that the entire resulting expression is built out of
+      congruence lemmas and `equiv.refl`, we use the `accept` subtactic of `solve_by_elim`
+      to reject any results which neither contain `eq` or a remaining metavariable.
+    * Since some congruence lemmas generate subgoals with `∀` statements,
+      we use the `pre_apply` subtactic of `solve_by_elim` to preprocess each new goal with `intros`.
+  -/
+  solve_by_elim {
+    use_symmetry := false,
+    use_exfalso := false,
+    lemmas := some (eq :: equiv_congr_lemmas),
+    -- TODO decide an appropriate upper bound on search depth.
```

Comment:
This should probably be configurable when you call `equiv_rw`.

Reply:
Good point, I will make this happen.

Reply:
Done.